Ensure weight tying is maintained for embed_tokens and lm_head

romitjain opened the pull request:
This PR solves: #2777 

Essentially, the PR:
1. Adds an argument in `LoraConfig` for users to maintain weight tying in the model
2. Checks if the passed model has weight tying enabled, gets the layers that are tied tothe  embedding layer
3. Add a `ModulesToSaveWrapper` on modules returned in (2) and a reference to the original layer

This ensures that the weight tying is enabled if the user passes either `embed_tokens` or `lm_head`.

Limitations
1. Only added for `LoraConfig`
2. Only added for `modules_to_save` parameter. This can be extended to `target_modules` as well once the approach is approved

BenjaminBossan commented on .gitignore:
```diff
@@ -143,3 +143,6 @@ wandb
 method_comparison/MetaMathQA/temporary_results/
+
+.peft/
+test.py
```
Let's remove these. You can set up a [global .gitignore](https://gist.github.com/subfuzion/db7f57fff2fb6998a16c) for these cases.

BenjaminBossan reviewed (commented):
Thanks for working on the PR to deal with targeting tied weights.

I have a suspicion that parts of this PR can be implemented more easily by extending the `peft_config.modules_to_save` to include the tied module instead of calling `_set_trainable` explicitly.

However, before checking that, it makes most sense to add one or a few unit tests to establish if everything works as expected. I think the best place for those would be in [this test class](https://github.com/huggingface/peft/blob/4469af57a08b0a751b7b1359d431c03a22b33116/tests/test_initialization.py#L76). LMK if you need help with implementing the tests.

Once we have the tests and confirm that everything works as expected, we can check if my proposal above works.

romitjain commented:
@BenjaminBossan I have addressed your comments. You mentioned that I can extend `peft_config.modules_to_save` to enable the functionality. Can you elaborate on that?

Essentially, we would need to mark the tied modules as modules to save and then update the weight pointer to point to the source adapter.

PTAL.
Thanks

BenjaminBossan commented on src/peft/peft_model.py:
```diff
@@ -1845,6 +1846,31 @@ def __init__(
+        # in `modules_to_save` and we want to ensure the `lm_head`
+        # does not diverge from the `embed_tokens` layer
+        if (
+            peft_config.task_type == "CAUSAL_LM"
```
Why is it important?

romitjain replied:
Yes, I'll make the changes such that we only worry about weight tying in case one of the tied layers is added in the modules to save

BenjaminBossan commented on tests/test_initialization.py:
```diff
@@ -1566,6 +1607,75 @@ def test_multiple_configs_with_bias_raises(self, tmp_path):
+            target_modules=["linear"],
+            ensure_weight_tieing=True,
+        )
+        model = get_peft_model(model, embed_token_config)
```
I think it would make sense to have a warning here, since it would seem that the user tried to tie the weights but there was nothing to tie, WDYT?

romitjain replied:
Yes, I have added a warning here:

https://github.com/huggingface/peft/pull/2803/files#diff-95787fcc5356568929e383449ade2c1bac5da43deccab17318652b62ed171ae7R1860

BenjaminBossan commented on .gitignore:
```diff
@@ -142,4 +142,4 @@ wandb
 # method_comparison logs
 method_comparison/MetaMathQA/cancelled_results/
-method_comparison/MetaMathQA/temporary_results/
+method_comparison/MetaMathQA/temporary_results/
```
Please add newline :)

romitjain commented:
@BenjaminBossan I think what you suggested is a cleaner approach. I am aligned with that.
The only concern I have is that we would first register the redundant weights (extra memory) for the tied modules and then replace the pointer post hoc.
However, the extra memory might not be too big, so it should be okay? What do you think?

BenjaminBossan commented on src/peft/tuners/tuners_utils.py:
```diff
@@ -1154,6 +1156,76 @@ def _get_tied_target_modules(self, model: nn.Module) -> list[str]:
+                warnings.warn("You have requested `ensure_weight_tying`, but no tied modules were found in the model")
+
+        else:
+            self._maybe_warn_ensure_weight_tying(is_embedding_to_save, tied_weight_keys)
```
I think this is a bit too specific as a method. How about removing `_maybe_warn_ensure_weight_tying` and just adding this:

```python
elif hasattr(peft_config, "ensure_weight_tying"):
    warnings.warn("Model has `tie_word_embeddings=True` ...")
```

romitjain commented on tests/test_initialization.py:
```diff
@@ -1566,6 +1600,93 @@ def test_multiple_configs_with_bias_raises(self, tmp_path):
         config2 = LoraConfig(target_modules=["linear"], bias="none")
         model.add_adapter("other", config2)  # does not raise
 
+    def test_weight_tying_tied_model(self):
```
Done @BenjaminBossan

githubnemo commented on src/peft/tuners/tuners_utils.py:
```diff
@@ -1154,6 +1156,86 @@ def _get_tied_target_modules(self, model: nn.Module) -> list[str]:
+        with any layers that needs to be tied
+        """
+        modules_to_save = set(getattr(peft_config, "modules_to_save", []) or [])
+        is_embedding_to_save = any(m in EMBEDDING_LAYER_NAMES for m in modules_to_save)
```
Should we use `model.get_input_embeddings` if it is available instead of guessing by name?

githubnemo replied:
makes sense. I think in the future we shouldn't rely on the embedding names if possible (e.g., custom models wouldn't be possible). but it's ok for me to leave it as is and do this change separately.

githubnemo reviewed (commented):
This change looks cool :)

One question, otherwise LGTM.
